### DISCUSSION TITLE
vps using layouts (renderer) setup

### DISCUSSION THREAD

**Discussion by Maikpwwq** [vikejs/vike#1081] [Discussion] [answered]
Hi. This is [my minimal reproduction project](https://github.com/Maikpwwq/vps-layouts), I am trying to build and app with different layouts according to his route, one for `index` and one for `app` folders, that why I have two `renderer` in my project. Project builds successfully. But in development mode I have troubles when navigating between layers. Please help me find the right way to perform that process.

**Comment by brillout** [comment]
> I have troubles when navigating between layers

What's the issue?

**Reply by Maikpwwq** [reply]
I travel to Profile clicking a btn that navigates to `http://localhost:3000/app/profile`, there loads my profile page OK. There I have a btn navigation pointing to Home `http://localhost:3000/`, I click then the URL changes but the page doesn't until I recharged everything.

In console fist navigation `app/profile` 

```
Warning: You are calling ReactDOMClient.createRoot() on a container that has already been passed to createRoot() before. Instead, call root.render() on the existing root instead if you want to update it.
```

In console second navigation `/`

```
Warning: render(...): It looks like the React-rendered content of the root container was removed without using React. This is not supported and will cause errors. Instead, call root.unmount() to empty a root's container.
```

**Reply by brillout** [reply]
Seems like a user land issue.

**Reply by Maikpwwq** [reply]
I was reviewing documentation about [domain-driven file structure](https://vite-plugin-ssr.com/file-structure#domain-driven) with multiple `pages/` directories, I need something like that so I decided to modify folder structure in my project, I watch your repository [file-structure-domain-driven](https://github.com/brillout/vite-plugin-ssr/tree/main/examples/file-structure-domain-driven). My question is this, if I need have more than one layout, the basic layout and at least another layout for a set of`pages/` directories, how can I use the `renderer` folder to achieve this? 

>  [my example repo](https://github.com/Maikpwwq/vps-layouts)

**Reply by brillout** [reply]
https://brillout.github.io/rules/#try. That said, let me know if you believe there is a (design) problem with vite-plugin-ssr that needs to be fixed.

**Reply by Maikpwwq** [reply]
Help me with this please. In my VPS project I first tried with single pages/ directory. Then when I implemented both Layers, I found some problems to continue. Now, I have update to Domain driven structure for VPS projects, but still I am facing this issue: ¿How to implement more the one layout for my pages with renderer folder setup??. For that reason I wrote here in [vite-plugin-ssr Help & Questions](https://github.com/brillout/vite-plugin-ssr/discussions/categories/help-questions) because the documentation was not enough to solve this, also I have build a minimal reproduction repository for create this discussions. Specifically in my files `_default.page.client.tsx` and `_default.page.server.tsx`, is been loaded `Page` from pageContext, inside an outer Layout, so I duplicated that structure where I use instead an inner layout, but it's been duplicated the root structure, Can we use your tools to achieve what we want?

**Reply by brillout** [reply]
I'm still inclined to think it's a user land issue. Feel free to provide a *minimal* reproduction showcasing your issue. ([This](https://github.com/Maikpwwq/vps-layouts/blob/db268b1f0e59e4223fb5f67fb7b3d2508c0fe363/package.json#L12-L41) isn't minimal.)

**Reply by Maikpwwq** [reply]
Excuse me @brillout but I don´t understand what you mean when you say `user land issue`. I am going to try with [Custom Export](https://vite-plugin-ssr.com/layouts#custom-export) to export and implement my multiple layouts. Thank you by the attendance.

I need to take a glance to this interesting conversation [578](https://github.com/brillout/vite-plugin-ssr/issues/578).

**Reply by brillout** [reply]
Basically, any issue that isn't related to vite-plugin-ssr's source code.

**Reply by Maikpwwq** [reply, marked as answer]
Ok, I got it. I was able to implement multiple layouts using [Custom Export](https://vite-plugin-ssr.com/layouts#custom-export) and [domain-driven file structure](https://vite-plugin-ssr.com/file-structure#domain-driven) in my project:

[vps layout's repo](https://github.com/Maikpwwq/vps-layouts)

Also, I had to change my file `tsconfig.json` to use correctly [ts-node esm support](https://www.npmjs.com/package/ts-node#native-ecmascript-modules). That because run package.json scripts with Node.js v20.1.0 were giving me this error in console:

```
node:internal/process/esm_loader:42
      internalBinding('errors').triggerUncaughtException(
                                ^
{ diagnosticCodes: [ 5110 ] }
```

This is how my file `tsconfig.json` looks like.

```
{
  "compilerOptions": {
    "strict": true,
    "module": "ESNext",
    "target": "ES2020",
    "moduleResolution": "Node",
    "lib": ["DOM", "DOM.Iterable", "ESNext"],
    "types": ["vite/client"],
    "jsx": "react-jsx",
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true
  },
  "ts-node": {
    "experimentalSpecifierResolution": "node",
    "transpileOnly": true,
    "esm": true,
    // This throws me that error 
    // "compilerOptions": {
    //   "moduleResolution": "Node16"
    // }
  }
}
```

**Reply by brillout** [reply]
So it was indeed a user land issue 😉. That's why I was mentioning https://brillout.github.io/rules/#try.

Thanks for answering your own question, hopefully it will help others.
> node:internal/process/esm_loader:42
>       internalBinding('errors').triggerUncaughtException(
>                                 ^
> { diagnosticCodes: [ 5110 ] }

See [this comment](https://github.com/brillout/vite-plugin-ssr/issues/1102#issuecomment-1701668469).
